File: packages/gateway/kotlin/gateway/src/main/kotlin/com/bridgething/gateway/LogStore.kt
package com.bridgething.gateway

import android.content.Context
import java.io.File
import java.io.IOException
import java.io.Writer
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.TimeUnit

public object LogStore {
    private const val LAUNCH_LIMIT = 3
    private const val SEGMENTS_PER_LAUNCH = 2
    private const val SEGMENT_BYTES = 512L * 1024
    private const val QUEUE_CAPACITY = 4096

    private const val PINNED_BYTES_LIMIT = 32L * 1024 * 1024
    private const val PIN_SUFFIX = ".keep"
    private val ERROR_LINE =
        Regex("""^\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\s+\d+\s+\d+\s+[EF]\s""")
    private val queue = ArrayBlockingQueue<String>(QUEUE_CAPACITY)
    private val ID = Regex("""^\d+$""")

    @Volatile private var root: File? = null
    @Volatile private var launchDir: File? = null
    @Volatile private var writer: Thread? = null
    @Volatile private var sinkGeneration = 0
    @Volatile private var dropped: Long = 0

    public fun install(context: Context) {
        install(File(context.applicationContext.filesDir, "logs"))
    }

    @Synchronized
    public fun install(dir: File) {
        if (writer != null) return

        dir.mkdirs()
        root = dir

        val launch = File(dir, System.currentTimeMillis().toString())
        launch.mkdirs()
        launchDir = launch
        pruneLaunches(dir)

        writer = Thread(::runWriter, "bridgething-logstore").apply {
            isDaemon = true
            priority = Thread.MIN_PRIORITY
            start()
        }
    }

    public fun write(line: String) {
        if (writer == null) return
        if (!queue.offer(line)) dropped += 1
    }

    public enum class Level(internal val letter: Char) {
        VERBOSE('V'),
        DEBUG('D'),
        INFO('I'),
        WARN('W'),
        ERROR('E'),
        FATAL('F'),
    }

    public fun record(level: Level, tag: String, message: String) {
        if (writer == null) return
        val prefix = "${stamp()} ${android.os.Process.myPid()} ${android.os.Process.myTid()} ${level.letter} $tag: "
        for (part in message.split('\n')) write(prefix + part)
    }

    private val STAMP = java.text.SimpleDateFormat("MM-dd HH:mm:ss.SSS", java.util.Locale.US)

    private fun stamp(): String = synchronized(STAMP) { STAMP.format(java.util.Date()) }

    // ---- export ----------------------------------------------------------

    public data class LogArchive(
        val id: String,
        val startedAtMs: Long,
        val bytes: Long,
        val pinned: Boolean,
        val current: Boolean,
    )

    public fun archives(): List<LogArchive> =
        launchDirs()
            .map { dir ->
                LogArchive(
                    id = dir.name,
                    startedAtMs = dir.name.toLongOrNull() ?: 0L,
                    bytes = segments(dir).sumOf { it.length() },
                    pinned = pinnedSegments(dir).isNotEmpty(),
                    current = dir == launchDir,
                )
            }
            .sortedByDescending { it.startedAtMs }

    @Synchronized
    public fun delete(id: String) {
        val base = root ?: return
        if (!ID.matches(id)) return
        val dir = File(base, id)
        if (!dir.isDirectory) return
        flush()
        if (dir == launchDir) {
            segments(dir).forEach { pinMarker(it).delete(); it.delete() }
            sinkGeneration += 1
        } else {
            dir.deleteRecursively()
        }
    }

    public fun exportTo(target: File, id: String? = null): File {
        flush()
        target.parentFile?.mkdirs()
        target.bufferedWriter().use { out ->
            out.write("bridgething log export\n")
            out.write("generated: ${java.util.Date()}\n")
            val dirs = if (id == null) launchDirs() else launchDirs().filter { it.name == id }
            out.write("launches: ${dirs.size}\n")
            if (dropped > 0) out.write("dropped lines (writer backpressure): $dropped\n")
            out.write("\n")
            for (dir in dirs) {
                val stamp = dir.name.toLongOrNull()?.let { java.util.Date(it) }?.toString() ?: dir.name
                val current = if (dir == launchDir) " (current)" else ""
                val pinned = if (pinnedSegments(dir).isNotEmpty()) " [pinned: contains errors]" else ""
                out.write("===== launch $stamp$current$pinned =====\n")
                for (segment in segments(dir)) {
                    runCatching { segment.forEachLine { out.write(it); out.write("\n") } }
                        .onFailure { out.write("<<unreadable segment ${segment.name}: ${it.message}>>\n") }
                }
                out.write("\n")
            }
        }
        return target
    }

    public fun retainedBytes(): Long = launchDirs().sumOf { dir -> segments(dir).sumOf { it.length() } }

    @Synchronized
    public fun clear() {
        flush()
        for (dir in launchDirs()) {
            if (dir == launchDir) {
                segments(dir).forEach { pinMarker(it).delete(); it.delete() }
            } else {
                dir.deleteRecursively()
            }
        }
        sinkGeneration += 1
        dropped = 0
    }

    public fun flush() {
        val deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(2)
        while (queue.isNotEmpty() && System.nanoTime() < deadline) Thread.sleep(10)
        Thread.sleep(50)
    }

    // ---- writer thread ---------------------------------------------------

    private fun runWriter() {
        var sink: Sink? = null
        var generation = sinkGeneration
        try {
            while (true) {
                val first = queue.take()
                if (generation != sinkGeneration) {
                    runCatching { sink?.close() }
                    sink = null
                    generation = sinkGeneration
                }
                var line: String? = first
                while (line != null) {
                    val active = sink ?: openSink().also { sink = it }
                    active.write(line)
                    if (active.bytes >= SEGMENT_BYTES) {
                        if (active.sawError) active.pin()
                        active.close()
                        sink = null
                    }
                    line = queue.poll()
                }
                sink?.let {
                    if (it.sawError) it.pin()
                    it.flush()
                }
            }
        } catch (_: InterruptedException) {
            // process teardown; nothing useful left to do
        } catch (e: IOException) {
            android.util.Log.w(TAG, "log writer stopped: ${e.message}")
        } finally {
            runCatching { sink?.close() }
        }
    }

    private fun openSink(): Sink {
        val dir = launchDir ?: throw IOException("log store not installed")
        val existing = segments(dir)
        val newest = existing.lastOrNull()
        val target = if (newest != null && newest.length() < SEGMENT_BYTES) {
            newest
        } else {
            val nextIndex = (newest?.let { segmentIndex(it) } ?: -1) + 1
            File(dir, "%04d.log".format(nextIndex))
        }
        pruneSegments(dir, keepAlso = target)
        return Sink(target)
    }

    private class Sink(private val file: File) {
        var bytes: Long = file.length()
            private set

        var sawError: Boolean = false
            private set

        private var pinned: Boolean = false

        private val out: Writer = java.io.OutputStreamWriter(
            java.io.FileOutputStream(file, true),
            Charsets.UTF_8,
        ).buffered()

        fun write(line: String) {
            out.write(line)
            out.write("\n")
            bytes += line.length.toLong() + 1
            if (!pinned && ERROR_LINE.containsMatchIn(line)) sawError = true
        }

        fun pin() {
            sawError = false
            if (pinned) return
            pinned = runCatching { pinMarker(file).createNewFile() || pinMarker(file).exists() }
                .getOrDefault(false)
        }

        fun flush() = out.flush()

        fun close() {
            runCatching { out.flush() }
            runCatching { out.close() }
        }
    }

    // ---- layout helpers --------------------------------------------------

    private fun launchDirs(): List<File> =
        root?.listFiles { f: File -> f.isDirectory }?.sortedBy { it.name } ?: emptyList()

    private fun segments(dir: File): List<File> =
        dir.listFiles { f: File -> f.isFile && f.name.endsWith(".log") }
            ?.sortedBy { segmentIndex(it) } ?: emptyList()

    private fun segmentIndex(file: File): Int = file.name.removeSuffix(".log").toIntOrNull() ?: -1

    private fun pinMarker(segment: File): File =
        File(segment.parentFile, segment.name.removeSuffix(".log") + PIN_SUFFIX)

    private fun isPinned(segment: File): Boolean = pinMarker(segment).exists()

    private fun pinnedSegments(dir: File): List<File> = segments(dir).filter { isPinned(it) }

    private fun pruneLaunches(dir: File) {
        val dirs = dir.listFiles { f: File -> f.isDirectory }?.sortedBy { it.name } ?: return
        val (pinned, rotating) = dirs.partition { pinnedSegments(it).isNotEmpty() }

        val excess = rotating.size - LAUNCH_LIMIT
        for (i in 0 until excess) rotating[i].deleteRecursively()

        var total = pinned.sumOf { d -> pinnedSegments(d).sumOf { it.length() } }
        for (d in pinned) {
            if (total <= PINNED_BYTES_LIMIT) break
            total -= pinnedSegments(d).sumOf { it.length() }
            android.util.Log.w(TAG, "pinned log cap hit; dropping error launch ${d.name}")
            d.deleteRecursively()
        }
    }

    private fun pruneSegments(dir: File, keepAlso: File) {
        val all = (segments(dir) + keepAlso).distinctBy { it.name }.sortedBy { segmentIndex(it) }
        val rotating = all.filterNot { isPinned(it) }
        val excess = rotating.size - SEGMENTS_PER_LAUNCH
        if (excess <= 0) return
        for (i in 0 until excess) if (rotating[i].name != keepAlso.name) rotating[i].delete()
    }

    private const val TAG = "bridgething-logs"
}

File: packages/gateway/swift/Sources/BridgethingGateway/LogStore.swift
import Foundation

public final class LogStore: @unchecked Sendable {
  public static let shared = LogStore()

  public enum Level: Character, Sendable {
    case trace = "V"
    case debug = "D"
    case info = "I"
    case notice = "N"
    case warn = "W"
    case error = "E"
    case fatal = "F"

    var pins: Bool { self == .error || self == .fatal }
  }

  public struct Limits: Sendable {
    public var launches: Int
    public var segmentsPerLaunch: Int
    public var segmentBytes: Int64
    public var queueCapacity: Int

    public var pinnedBytesLimit: Int64

    public init(
      launches: Int = 3,
      segmentsPerLaunch: Int = 2,
      segmentBytes: Int64 = 512 * 1024,
      queueCapacity: Int = 4096,
      pinnedBytesLimit: Int64 = 32 * 1024 * 1024
    ) {
      self.launches = launches
      self.segmentsPerLaunch = segmentsPerLaunch
      self.segmentBytes = segmentBytes
      self.queueCapacity = queueCapacity
      self.pinnedBytesLimit = pinnedBytesLimit
    }
  }

  public struct Archive: Sendable, Equatable {
    public let id: String
    public let startedAtMs: Int64
    public let bytes: Int64
    public let pinned: Bool
    public let current: Bool
  }

  private struct Entry {
    let line: String
    let pins: Bool
  }

  private static let pinSuffix = ".keep"
  private static let segmentSuffix = ".log"

  private let limits: Limits
  private let fm = FileManager.default

  private let queue = NSCondition()
  private var pending: [Entry] = []
  private var running = false
  private var draining = false

  private var dropped: Int64 = 0

  private var sinkGeneration = 0

  private let state = NSLock()
  private var root: URL?
  private var launchName: String?
  private var thread: Thread?

  public init(limits: Limits = Limits()) {
    self.limits = limits
  }

  // MARK: - install

  public static func defaultRoot() -> URL {
    let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
      ?? FileManager.default.temporaryDirectory
    return base.appendingPathComponent("bridgething-logs", isDirectory: true)
  }

  public func install(root url: URL = LogStore.defaultRoot()) {
    state.lock()
    defer { state.unlock() }
    guard thread == nil else { return }

    try? fm.createDirectory(at: url, withIntermediateDirectories: true)
    root = url

    let name = String(Int64(Date().timeIntervalSince1970 * 1000))
    let launch = url.appendingPathComponent(name, isDirectory: true)
    try? fm.createDirectory(at: launch, withIntermediateDirectories: true)
    launchName = name
    pruneLaunches(in: url)

    queue.lock()
    running = true
    queue.unlock()

    let worker = Thread { [weak self] in self?.runWriter(launch: launch) }
    worker.name = "bridgething-logstore"
    worker.qualityOfService = .background
    thread = worker
    worker.start()
  }

  // MARK: - ingest

  public func record(level: Level, label: String, message: String) {
    queue.lock()
    let live = running
    queue.unlock()
    guard live else { return }

    let prefix = "\(Self.stamp(Date())) \(Self.pid) \(Self.currentThreadId()) \(level.rawValue) \(label): "
    queue.lock()
    defer { queue.unlock() }
    for part in message.split(separator: "\n", omittingEmptySubsequences: false) {
      if pending.count >= limits.queueCapacity {
        dropped += 1
        continue
      }
      pending.append(Entry(line: prefix + part, pins: level.pins))
    }
    queue.broadcast()
  }

  // MARK: - export

  public func archives() -> [Archive] {
    let (base, live) = currentLayout()
    guard let base else { return [] }
    return Self.launchDirs(in: base)
      .map { dir in
        Archive(
          id: dir.lastPathComponent,
          startedAtMs: Int64(dir.lastPathComponent) ?? 0,
          bytes: Self.totalBytes(Self.segments(in: dir)),
          pinned: !Self.pinnedSegments(in: dir).isEmpty,
          current: dir.lastPathComponent == live
        )
      }
      .sorted { $0.startedAtMs > $1.startedAtMs }
  }

  public func retainedBytes() -> Int64 {
    let (base, _) = currentLayout()
    guard let base else { return 0 }
    return Self.launchDirs(in: base).reduce(0) { $0 + Self.totalBytes(Self.segments(in: $1)) }
  }

  public func delete(id: String) {
    state.lock()
    defer { state.unlock() }
    guard let base = root, Self.isLaunchId(id) else { return }
    let dir = base.appendingPathComponent(id, isDirectory: true)
    guard Self.isDirectory(dir) else { return }
    flush()
    if id == launchName {
      truncate(dir)
      bumpGeneration()
    } else {
      try? fm.removeItem(at: dir)
    }
  }

  public func clear() {
    state.lock()
    defer { state.unlock() }
    guard let base = root else { return }
    flush()
    for dir in Self.launchDirs(in: base) {
      if dir.lastPathComponent == launchName {
        truncate(dir)
      } else {
        try? fm.removeItem(at: dir)
      }
    }
    bumpGeneration()
    queue.lock()
    dropped = 0
    queue.unlock()
  }

  @discardableResult
  public func exportTo(_ target: URL, id: String? = nil) throws -> URL {
    flush()
    try fm.createDirectory(at: target.deletingLastPathComponent(), withIntermediateDirectories: true)
    fm.createFile(atPath: target.path, contents: nil)
    let out = try FileHandle(forWritingTo: target)
    defer { try? out.close() }

    let (base, live) = currentLayout()
    var dirs = base.map { Self.launchDirs(in: $0) } ?? []
    if let id { dirs = dirs.filter { $0.lastPathComponent == id } }

    queue.lock()
    let lost = dropped
    queue.unlock()

    var header = "bridgething log export\n"
    header += "generated: \(Date())\n"
    header += "launches: \(dirs.count)\n"
    if lost > 0 { header += "dropped lines (writer backpressure): \(lost)\n" }
    header += "\n"
    try out.write(contentsOf: Data(header.utf8))

    for dir in dirs {
      let name = dir.lastPathComponent
      let stamp = Int64(name).map { "\(Date(timeIntervalSince1970: Double($0) / 1000))" } ?? name
      let current = name == live ? " (current)" : ""
      let pinned = Self.pinnedSegments(in: dir).isEmpty ? "" : " [pinned: contains errors]"
      try out.write(contentsOf: Data("===== launch \(stamp)\(current)\(pinned) =====\n".utf8))
      for segment in Self.segments(in: dir) {
        do {
          let body = try Data(contentsOf: segment, options: .mappedIfSafe)
          try out.write(contentsOf: body)
          if body.last != 0x0A { try out.write(contentsOf: Data([0x0A])) }
        } catch {
          let note = "<<unreadable segment \(segment.lastPathComponent): \(error.localizedDescription)>>\n"
          try out.write(contentsOf: Data(note.utf8))
        }
      }
      try out.write(contentsOf: Data("\n".utf8))
    }
    return target
  }

  public func flush() {
    queue.lock()
    defer { queue.unlock() }
    let deadline = Date().addingTimeInterval(2)
    while !pending.isEmpty || draining {
      guard Date() < deadline else { return }
      queue.wait(until: deadline)
    }
  }

  // MARK: - writer thread

  private func runWriter(launch: URL) {
    var sink: Sink?
    var generation = 0
    while true {
      queue.lock()
      while pending.isEmpty { queue.wait() }
      let batch = pending
      pending.removeAll(keepingCapacity: true)
      draining = true
      let wanted = sinkGeneration
      queue.unlock()

      if wanted != generation {
        sink?.close()
        sink = nil
        generation = wanted
      }

      do {
        for entry in batch {
          let active: Sink
          if let existing = sink {
            active = existing
          } else {
            active = try openSink(in: launch)
            sink = active
          }
          active.write(entry.line, pins: entry.pins)
          if active.bytes >= limits.segmentBytes {
            if active.sawError { active.pin() }
            active.close()
            sink = nil
          }
        }
        if let active = sink {
          if active.sawError { active.pin() }
          try active.flush()
        }
      } catch {
        sink?.close()
        sink = nil
      }

      queue.lock()
      draining = false
      queue.broadcast()
      queue.unlock()
    }
  }

  private func openSink(in launch: URL) throws -> Sink {
    let existing = Self.segments(in: launch)
    let newest = existing.last
    let target: URL
    if let newest, Self.fileSize(newest) < limits.segmentBytes {
      target = newest
    } else {
      let next = (newest.map { Self.segmentIndex($0) } ?? -1) + 1
      target = launch.appendingPathComponent(String(format: "%04d", next) + Self.segmentSuffix)
    }
    pruneSegments(in: launch, keepAlso: target)
    return try Sink(url: target)
  }

  private final class Sink {
    private(set) var bytes: Int64

    private(set) var sawError = false

    private let url: URL
    private let handle: FileHandle
    private var pinned = false
    private var buffer = Data()

    init(url: URL) throws {
      self.url = url
      if !FileManager.default.fileExists(atPath: url.path) {
        FileManager.default.createFile(atPath: url.path, contents: nil)
      }
      handle = try FileHandle(forWritingTo: url)
      let end = try handle.seekToEnd()
      bytes = Int64(end)
    }

    func write(_ line: String, pins: Bool) {
      let utf8 = Data(line.utf8)
      buffer.append(utf8)
      buffer.append(0x0A)
      bytes += Int64(utf8.count + 1)
      if !pinned, pins { sawError = true }
    }

    func pin() {
      sawError = false
      guard !pinned else { return }
      let marker = LogStore.pinMarker(for: url)
      pinned = (try? Data().write(to: marker)) != nil || FileManager.default.fileExists(atPath: marker.path)
    }

    func flush() throws {
      guard !buffer.isEmpty else { return }
      try handle.write(contentsOf: buffer)
      buffer.removeAll(keepingCapacity: true)
    }

    func close() {
      try? flush()
      try? handle.close()
    }
  }

  // MARK: - rotation

  private func pruneLaunches(in base: URL) {
    let dirs = Self.launchDirs(in: base)
    let pinned = dirs.filter { !Self.pinnedSegments(in: $0).isEmpty }
    let rotating = dirs.filter { Self.pinnedSegments(in: $0).isEmpty }

    let excess = rotating.count - limits.launches
    if excess > 0 {
      for dir in rotating.prefix(excess) { try? fm.removeItem(at: dir) }
    }

    var total = pinned.reduce(Int64(0)) { $0 + Self.totalBytes(Self.pinnedSegments(in: $1)) }
    for dir in pinned {
      if total <= limits.pinnedBytesLimit { break }
      total -= Self.totalBytes(Self.pinnedSegments(in: dir))
      try? fm.removeItem(at: dir)
    }
  }

  private func pruneSegments(in launch: URL, keepAlso: URL) {
    var all = Self.segments(in: launch)
    if !all.contains(where: { $0.lastPathComponent == keepAlso.lastPathComponent }) { all.append(keepAlso) }
    all.sort { Self.segmentIndex($0) < Self.segmentIndex($1) }
    let rotating = all.filter { !Self.isPinned($0) }
    let excess = rotating.count - limits.segmentsPerLaunch
    guard excess > 0 else { return }
    for dir in rotating.prefix(excess) where dir.lastPathComponent != keepAlso.lastPathComponent {
      try? fm.removeItem(at: dir)
    }
  }

  private func truncate(_ dir: URL) {
    for segment in Self.segments(in: dir) {
      try? fm.removeItem(at: Self.pinMarker(for: segment))
      try? fm.removeItem(at: segment)
    }
  }

  private func bumpGeneration() {
    queue.lock()
    sinkGeneration += 1
    queue.unlock()
  }

  private func currentLayout() -> (URL?, String?) {
    state.lock()
    defer { state.unlock() }
    return (root, launchName)
  }

  // MARK: - layout helpers

  private static func isLaunchId(_ id: String) -> Bool {
    !id.isEmpty && id.allSatisfy(\.isASCII) && id.allSatisfy(\.isNumber)
  }

  private static func isDirectory(_ url: URL) -> Bool {
    var isDir: ObjCBool = false
    return FileManager.default.fileExists(atPath: url.path, isDirectory: &isDir) && isDir.boolValue
  }

  private static func launchDirs(in base: URL) -> [URL] {
    let entries = (try? FileManager.default.contentsOfDirectory(at: base, includingPropertiesForKeys: nil)) ?? []
    return entries
      .filter { isLaunchId($0.lastPathComponent) && isDirectory($0) }
      .sorted { $0.lastPathComponent < $1.lastPathComponent }
  }

  private static func segments(in dir: URL) -> [URL] {
    let entries = (try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)) ?? []
    return entries
      .filter { $0.lastPathComponent.hasSuffix(segmentSuffix) }
      .sorted { segmentIndex($0) < segmentIndex($1) }
  }

  private static func segmentIndex(_ url: URL) -> Int {
    Int(url.lastPathComponent.replacingOccurrences(of: segmentSuffix, with: "")) ?? -1
  }

  private static func pinMarker(for segment: URL) -> URL {
    let base = segment.lastPathComponent.replacingOccurrences(of: segmentSuffix, with: "")
    return segment.deletingLastPathComponent().appendingPathComponent(base + pinSuffix)
  }

  private static func isPinned(_ segment: URL) -> Bool {
    FileManager.default.fileExists(atPath: pinMarker(for: segment).path)
  }

  private static func pinnedSegments(in dir: URL) -> [URL] {
    segments(in: dir).filter { isPinned($0) }
  }

  private static func fileSize(_ url: URL) -> Int64 {
    let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
    return (attrs?[.size] as? NSNumber)?.int64Value ?? 0
  }

  private static func totalBytes(_ urls: [URL]) -> Int64 {
    urls.reduce(0) { $0 + fileSize($1) }
  }

  // MARK: - line prefix

  private static let pid = ProcessInfo.processInfo.processIdentifier

  private static func currentThreadId() -> UInt64 {
    var id: UInt64 = 0
    pthread_threadid_np(nil, &id)
    return id
  }

  private static let stampLock = NSLock()
  private static let stampFormatter: DateFormatter = {
    let f = DateFormatter()
    f.locale = Locale(identifier: "en_US_POSIX")
    f.dateFormat = "MM-dd HH:mm:ss.SSS"
    return f
  }()

  static func stamp(_ date: Date, timeZone: TimeZone? = nil) -> String {
    stampLock.lock()
    defer { stampLock.unlock() }
    let previous = stampFormatter.timeZone
    if let timeZone { stampFormatter.timeZone = timeZone }
    defer { stampFormatter.timeZone = previous }
    return stampFormatter.string(from: date)
  }
}
